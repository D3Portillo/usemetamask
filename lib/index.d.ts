import {
  Metamask,
  RequestFn,
  SendMethodProps,
  UseMatamaskAPI,
} from "./shared.d"

interface AddEtherNetwork {
  chainName: string
  /** 0x prefixed Network id */
  chainId: string
  rpcUrl: string
  blockExplorerUrl?: string
  reloadOnSuccess?: boolean
  currencySymbol?: string
}

interface AddEtherToken {
  address: string
  symbol: string
  decimals?: number
  image?: string
}

export function useMetamask(): UseMatamaskAPI
export function addEtherNetwork(props: AddEtherNetwork): Promise<null>
export function addEtherToken(props: AddEtherToken): Promise<null>
export function getMetamaskProvider(): Metamask | null
export function switchEtherNetwork(chainId: string): Promise<null>
export function switchOrAppendNetwork(props: AddEtherNetwork): Promise<null>
/**
 * Send ether transaction
 * @returns Transaction hash string
 */
export function etherSend(props: SendMethodProps): Promise<string>
export const etherRequest: RequestFn
export const parse = {
  toHex: (n: number): string => null,
  toInt: (s: string): string => null,
}
