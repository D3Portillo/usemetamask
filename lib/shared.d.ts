// * Library shared type definitions
// REF: https://docs.metamask.io
interface ErrorState {
  message: string
  code: number
}

interface ProviderMessage {
  type: string
  data: unknown
}

interface ProviderRpcError {
  message: string
  code: number
  data?: unknown
}

interface ConnectInfo {
  chainId: string
}

interface SendBasicProps {
  /** Address to send */
  to: string
  /** Hex formated wei amount to send */
  value: string
}

interface SendMethodProps extends SendBasicProps {
  gasPrice?: string
  /** Optional. Metamask UI will be prompted */
  gas?: string
  from: string
  data?: string
  /** Optional. Metamask will replace it's value */
  chainId?: string
}

interface Asset {
  address: string
  symbol: string
  decimals: number
  image: string
}

interface WatchAssetParams {
  type: "ERC20"
  options: Asset
}

interface AddEthereumChain {
  chainId: string
  chainName: string
  nativeCurrency?: {
    name: string
    symbol: string
    decimals: 18
  }
  rpcUrls: string[]
  blockExplorerUrls?: string[]
}

interface SwitchEthereumChain {
  chainId: string
}

// * metamask.request() typing
interface RequestProps<T, D> {
  method: T
  params: D
}

type TransactionType = "latest" | "earliest" | "pending"
interface OnRequest {
  request(
    p: RequestProps<"wallet_watchAsset", WatchAssetParams>
  ): Promise<boolean>
  request(
    p: RequestProps<"wallet_switchEthereumChain", SwitchEthereumChain[]>
  ): Promise<null>
  request(
    p: RequestProps<"wallet_addEthereumChain", AddEthereumChain[]>
  ): Promise<null>
  request(p: { method: "eth_requestAccounts" }): Promise<string[]>
  request(p: { method: "eth_accounts" }): Promise<[string][]>
  request(
    p: RequestProps<"eth_sendTransaction", SendMethodProps[]>
  ): Promise<string[]>
  /** Get Address ETH balance in wei */
  request(
    p: RequestProps<"eth_getBalance", [string, TransactionType]>
  ): Promise<string>
}

// * metamask.on() typing
type EventCallback<T> = (data: T) => void
interface OnEvent {
  on(ev: "accountsChanged", cb: EventCallback<string[]>): void
  on(ev: "chainChanged", cb: EventCallback<string>): void
  on(ev: "disconnect", cb: EventCallback<ProviderRpcError>): void
  on(ev: "message", cb: EventCallback<ProviderMessage>): void
  on(ev: "connect", cb: EventCallback<ConnectInfo>): void
}

// * Exported types
export type FIATCurrencies = {
  usd: number
  jpy: number
  eur: number
}

export interface Metamask extends OnEvent, OnRequest {
  isConnected(): boolean
  /**
   * alias to `ethereum._metamask.isUnlocked()`
   *
   * This method returns a Promise that resolves to a boolean indicating
   * if MetaMask is unlocked by the user.
   */
  isUserUnlocked(): Promise<boolean>
  isMetaMask: boolean
  chainId: string
  removeListener: (e: string, f: any) => void
  selectedAddress: string | null
}

export interface UseMatamaskAPI {
  connect: () => void
  disconnect: (props?: { reload?: boolean }) => void
  send: (props: Omit<SendMethodProps, "from">) => Promise<string>
  resetError: () => void
  account: string
  accounts: string[]
  balance: number
  formattedBalance: string
  chainId: string
  chainIdDecimal: number
  error?: ErrorState
  metamask: Metamask
  isConnected: boolean
}
