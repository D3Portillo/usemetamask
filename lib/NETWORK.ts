const Mainnet = {
  chainId: "0x1",
  chainName: "Ethereum Main Network (Mainnet)",
  rpcUrl: "https://mainnet.infura.io/v3/",
  blockExplorerUrl: "https://etherscan.io",
}

const Rinkeby = {
  chainId: "0x4",
  chainName: "Rinkeby Test Network",
  rpcUrl: "https://rinkeby.infura.io/v3/",
  blockExplorerUrl: "https://rinkeby.etherscan.io",
}

const Goerli = {
  chainId: "0x5",
  chainName: "Goerli Test Network",
  rpcUrl: "https://goerli.infura.io/v3/",
  blockExplorerUrl: "https://goerli.etherscan.io",
}

const Kovan = {
  chainId: "0x2a",
  chainName: "Kovan Test Network",
  rpcUrl: "https://kovan.infura.io/v3/",
  blockExplorerUrl: "https://kovan.etherscan.io",
}

const Ropsten = {
  chainId: "0x3",
  chainName: "Ropsten Test Network",
  rpcUrl: "https://ropsten.infura.io/v3/",
  blockExplorerUrl: "https://ropsten.etherscan.io",
}

const Matic = {
  chainId: "0x89",
  chainName: "Matic Mainnet",
  rpcUrl: "https://rpc-mainnet.maticvigil.com",
  blockExplorerUrl: "https://explorer.matic.network",
  symbol: "MATIC",
}

export default {
  Ropsten,
  Mainnet,
  Rinkeby,
  Goerli,
  Kovan,
  Matic,
}
