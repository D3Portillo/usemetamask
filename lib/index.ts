export { default as useMetamask } from "./useMetamask"
export { default as useTokenPrice } from "./useTokenPrice"
export {
  addEtherNetwork,
  addEtherToken,
  getMetamaskProvider,
  switchEtherNetwork,
  switchOrAppendNetwork,
  metamaskRequest,
  sendEther,
  parse,
  connectToMetamask,
  getFIATBalance,
  formatEther,
} from "./utils"
