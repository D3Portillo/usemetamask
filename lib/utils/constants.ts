export const METHODS = {
  WALLET_WATCH_ASSET: "wallet_watchAsset",
  WALLET_SWITCH_ETHER_CHAIN: "wallet_switchEthereumChain",
  WALLET_ADD_ETHER_CHAIN: "wallet_addEthereumChain",
  ETH_REQUEST_ACCOUNTS: "eth_requestAccounts",
  ETH_SEND_TRANSACTION: "eth_sendTransaction",
  ETH_ACCOUNTS: "eth_accounts",
} as const

export const ERRORS = {
  METAMASK_NOT_INSTALLED: {
    code: 333,
    message: "Metamask not installed",
  },
  USER_NOT_CONNECTED: {
    code: 0,
    message: "User not connected",
  },
} as const

export const EVENTS = {
  ON_METAMASK_ERROR: "ON_METAMASK_ERROR",
} as const

export const STR_TRUE = "true"
export const STR_FALSE = "false"
