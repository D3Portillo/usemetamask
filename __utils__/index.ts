import { act } from "@testing-library/react"
import { Metamask } from "../lib/shared"

export const waitForUseEffect = async (node) => {
  return act(async () => {
    await new Promise((resolve) => resolve(node))
  })
}

export function exposeMetamask(metamaskEth: {
  request?: Metamask["request"]
  on?: Metamask["on"]
}) {
  window["ethereum"] = {
    selectedAddress: "ADDRESS",
    isMetaMask: true,
    removeListener: jest.fn(),
    on: jest.fn(),
    ...metamaskEth,
  }
}
