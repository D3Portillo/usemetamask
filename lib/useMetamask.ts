import { useEffect, useState } from "react"
import { Metamask, SendBasicProps, UseMatamaskAPI } from "./shared.d"
import { getMetamaskProvider, etherSend } from "./utils"
import {
  setStoreState,
  userIsForceDisconnected,
  connectToMetamask,
} from "./utils/helpers"
import { STR_FALSE, STR_TRUE, EVENTS } from "./utils/constants"

const ACCOUNTS_CHANGED = "accountsChanged"
export function useMetamask(): UseMatamaskAPI {
  const [error, setError] = useState(null)
  const [metamask, setMetamask] = useState<Metamask>({} as any)
  const [accounts, setAccounts] = useState([])
  const [account, setAccount] = useState(null)

  const handleError = (err) => {
    console.error(`useMetamask::Error`, err)
    setError(err)
  }

  useEffect(() => {
    const metamask = getMetamaskProvider()
    function handleEventError(e) {
      e.detail && handleError(e.detail)
    }
    if (metamask) {
      setMetamask(metamask)
      metamask.on(ACCOUNTS_CHANGED, setAccounts)
      metamask.request({ method: "eth_accounts" }).then(setAccounts)
    }
    window.addEventListener(EVENTS.ON_METAMASK_ERROR, handleEventError)
    return () => {
      metamask && metamask.removeListener(ACCOUNTS_CHANGED, setAccounts)
      window.removeEventListener(EVENTS.ON_METAMASK_ERROR, handleEventError)
    }
  }, [])

  useEffect(() => {
    const isForceDisconnected = userIsForceDisconnected()
    const userCanConnect = accounts.length && !isForceDisconnected
    setAccount(userCanConnect ? accounts[0] : null)
  }, [accounts])

  useEffect(() => {
    // Error clean up
    setError(null)
  }, [accounts, account])

  const proxiedSend = ({ to, value }: SendBasicProps) => {
    return etherSend({
      from: account,
      to,
      value,
    })
  }

  const proxiedConnect = () => {
    connectToMetamask()
      .then((account) => {
        setStoreState(STR_FALSE)
        setAccounts([account])
      })
      .catch(handleError)
  }

  const disconnect = (__props) => {
    const props = __props || {}
    setStoreState(STR_TRUE)
    setAccounts([])
    props.reload && window.location.reload()
  }

  const chainId = metamask.chainId ? metamask.chainId : "0x0"
  return {
    send: proxiedSend,
    isConnected: typeof account === "string",
    chainIdDecimal: parseInt(chainId),
    chainId,
    metamask,
    connect: proxiedConnect,
    disconnect,
    account,
    error,
    accounts,
  }
}

export default useMetamask
