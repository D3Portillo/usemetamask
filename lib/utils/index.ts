/**
 * Library exported utilities
 */
import { METHODS } from "./constants"
import { Asset, SendBasicProps } from "../shared.d"
import { withError, runIfMetamask, getMetamaskProvider } from "./helpers"

export { getMetamaskProvider }

export const addEtherToken = ({
  address,
  symbol,
  decimals = 18,
  image = "",
}: Omit<Asset, "decimals" | "image"> & {
  decimals?: number
  image?: string
}) => {
  return runIfMetamask((metamask) => {
    return metamask
      .request({
        method: METHODS.WALLET_WATCH_ASSET,
        params: {
          type: "ERC20",
          options: {
            address,
            symbol,
            decimals,
            image,
          },
        },
      })
      .catch(withError)
  })
}

export const switchOrAppendNetwork = ({
  chainId,
  chainName,
  rpcUrl,
  reloadOnSuccess,
  blockExplorerUrl,
  currencySymbol = "ETH",
}): Promise<null> => {
  return runIfMetamask(async (metamask) => {
    return metamask
      .request({
        method: METHODS.WALLET_SWITCH_ETHER_CHAIN,
        params: [{ chainId }],
      })
      .catch((error) => {
        if (error.code === 4902) {
          return addEtherNetwork({
            chainName,
            chainId,
            rpcUrl,
            reloadOnSuccess,
            blockExplorerUrl,
            currencySymbol,
          })
        }
        return withError(error)
      })
  })
}

export const switchEtherNetwork = (chainId: string): Promise<null> => {
  return runIfMetamask((metamask) => {
    return metamask
      .request({
        method: METHODS.WALLET_SWITCH_ETHER_CHAIN,
        params: [{ chainId }],
      })
      .catch(withError)
  })
}

export const addEtherNetwork = ({
  chainName,
  chainId,
  rpcUrl,
  reloadOnSuccess,
  blockExplorerUrl,
  currencySymbol = "ETH",
}): Promise<null> => {
  return runIfMetamask((metamask) => {
    metamask
      .request({
        method: METHODS.WALLET_ADD_ETHER_CHAIN,
        params: [
          {
            chainName,
            chainId,
            rpcUrls: [rpcUrl],
            blockExplorerUrls: [blockExplorerUrl],
            nativeCurrency: {
              name: currencySymbol,
              symbol: currencySymbol,
              decimals: 18,
            },
          },
        ],
      })
      .then(() => {
        reloadOnSuccess && window.location.reload()
      })
      .catch(withError)
  })
}

export const metamaskRequest = (props) => {
  return runIfMetamask((metamask) => {
    return metamask.request(props).catch(withError)
  })
}

export const sendEther = ({
  from,
  to,
  value,
}: SendBasicProps & { from: string }): Promise<string> => {
  return runIfMetamask((metamask) => {
    return metamask
      .request({
        method: METHODS.ETH_SEND_TRANSACTION,
        params: [
          {
            from,
            to,
            value,
          },
        ],
      })
      .catch(withError)
  })
}

export const parse = {
  toHex: (n: number) => {
    if (n < 0) {
      // Handle negative overflow
      n = 0xffffffff + n + 1
    }
    return `0x${n.toString(16).toUpperCase()}`
  },
  hexToInt: (str: string) => parseInt(str, 16),
  toWei: (n: number) => n * 10 ** 18,
  toTxWei: (n: number) => parse.toHex(parse.toWei(n)),
  weiToEth: (n: number) => n * 10 ** -18,
  txWeiToEth: (s: string) => parse.weiToEth(parse.hexToInt(s)),
}
