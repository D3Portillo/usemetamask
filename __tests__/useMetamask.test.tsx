import * as React from "react"
import { act, render, RenderResult } from "@testing-library/react"

import { ERRORS } from "../lib/utils/constants"
import { UseMatamaskAPI } from "../lib/shared"
import { useMetamask } from "../lib/index"

import { ACCOUNTS } from "./__utils__/constans"
import { exposeMetamask, waitForUseEffect } from "./__utils__"

function ExposeHook({
  onRender = (_: UseMatamaskAPI) => {},
  onMetamask = undefined as any,
}) {
  onRender(useMetamask(onMetamask))
  return null
}

describe("useMetamask", () => {
  let node: RenderResult
  afterEach(() => {
    if (node && node.unmount) {
      node.unmount()
    }
    jest.clearAllMocks()
  })

  it("should match empty state & no metamask installed", async () => {
    jest.useFakeTimers()
    const expectedState = {
      chainIdDecimal: 0,
      chainId: "0x0",
      metamask: {},
      formattedBalance: "0.00",
      balance: 0,
      error: ERRORS.METAMASK_NOT_INSTALLED,
    }
    const onRender = jest.fn()
    const consoleError = jest
      .spyOn(console, "error")
      .mockImplementation(() => {})
    node = render(<ExposeHook onRender={onRender} />)
    expect(node.container).toMatchSnapshot()
    expect(consoleError).toHaveBeenCalled()
    act(() => {
      jest.runAllTimers()
    })
    expect(onRender).toBeCalledWith(expect.objectContaining(expectedState))
    consoleError.mockRestore()
  })

  it("should render account='0x111'", async () => {
    const expectedState = {
      account: ACCOUNTS[0],
      accounts: ACCOUNTS,
    }
    exposeMetamask({
      request: jest.fn().mockResolvedValue(ACCOUNTS),
    })
    const mockOnRender = jest.fn()
    node = render(<ExposeHook onRender={mockOnRender} />)
    await waitForUseEffect(node)
    expect(mockOnRender).toBeCalledWith(expect.objectContaining(expectedState))
  })

  it("should disconnect when connected", async () => {
    let api: UseMatamaskAPI = {} as any
    const expectedState = {
      isConnected: true,
    }
    exposeMetamask({
      request: jest.fn().mockResolvedValue(ACCOUNTS),
    })
    const mockOnRender = jest.fn((_api) => {
      api = _api
    })
    node = render(<ExposeHook onRender={mockOnRender} />)
    await waitForUseEffect(node)
    expect(mockOnRender).toBeCalledWith(expect.objectContaining(expectedState))
    expect(api.isConnected).toBeTruthy()
    act(() => {
      api.disconnect()
    })
    expect(api.isConnected).toBeFalsy()
  })

  it("should expose metamask at hook mount & execute cleanerFn", async () => {
    const exposedMetamask = exposeMetamask({
      request: jest.fn().mockResolvedValue(ACCOUNTS),
    })
    const mockOnRender = jest.fn()
    const mockCleanerFn = jest.fn()
    const mockOnMetamask = jest.fn(() => mockCleanerFn)
    node = render(
      <ExposeHook onRender={mockOnRender} onMetamask={mockOnMetamask} />
    )
    await waitForUseEffect(node)
    expect(mockOnMetamask).toBeCalledWith(exposedMetamask)
    node.unmount() // Remove component to trigger hookCleaner
    expect(mockCleanerFn).toHaveBeenCalledTimes(1)
  })
})
