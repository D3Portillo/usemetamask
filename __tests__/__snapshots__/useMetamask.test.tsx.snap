// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useMetamask should match empty state & no metamask installed 1`] = `<div />`;
